cdap-ui/app/cdap/components/DataPrep/Directives/MergeColumns/index.js

```diff
@@ -0,0 +1,302 @@
+ * the License.
+ */
+
+import React, { Component, PropTypes } from 'react';
+import classnames from 'classnames';
+import {execute} from 'components/DataPrep/store/DataPrepActionCreator';
+import DataPrepStore from 'components/DataPrep/store';
+import DataPrepActions from 'components/DataPrep/store/DataPrepActions';
+import WarningContainer from 'components/WarningContainer';
+import T from 'i18n-react';
+require('./MergeColumns.scss');
+
+const PREFIX = `features.DataPrep.Directives.Merge`;
+const DEFAULT_DELIMITER = 'COMMA';
+const DELIMITER_MAP = {
+  'COMMA': ',',
+  'PIPE': '|',
+  'COLON': ':',
+  'PERIOD': '.',
+  'DASH': '-',
+  'UNDERSCORE': '_'
+};
+
+export default class MergeColumnsDirective extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      firstColumn: this.props.column[0],
+      secondColumn: this.props.column[1],
+      newColumnInput: `${this.props.column[0]}_${this.props.column[1]}`,
+      selectedDelimiter: DEFAULT_DELIMITER,
+      customDelimiter: '',
+      isOpen: this.props.isOpen
+    };
+
+    this.handleNewColumnInputChange = this.handleNewColumnInputChange.bind(this);
+    this.handleDelimiterSelect = this.handleDelimiterSelect.bind(this);
+    this.handleCustomDelimiterChange = this.handleCustomDelimiterChange.bind(this);
+    this.switchColumnOrder = this.switchColumnOrder.bind(this);
+    this.handleKeyPress = this.handleKeyPress.bind(this);
+    this.applyDirective = this.applyDirective.bind(this);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.isOpen !== this.state.isOpen) {
+      this.setState({
+        isOpen: nextProps.isOpen
+      }, () => {
+        if (this.newColumnInputBox) {
+          this.newColumnInputBox.focus();
+        }
+      });
+    }
+  }
+
+  componentDidUpdate() {
+    if (this.state.selectedDelimiter.substr(0, 6) === 'CUSTOM' && this.state.customDelimiter.length === 0 && this.customDelimiterRef) {
+      this.customDelimiterRef.focus();
+    }
+  }
+
+  preventPropagation(e) {
+    e.stopPropagation();
+    e.nativeEvent.stopImmediatePropagation();
+    e.preventDefault();
+  }
+
+  handleNewColumnInputChange(e) {
+    this.setState({newColumnInput: e.target.value});
+  }
+
+  handleDelimiterSelect(e) {
+    this.setState({selectedDelimiter: e.target.value});
+  }
+
+  handleCustomDelimiterChange(e) {
+    this.setState({customDelimiter: e.target.value});
+  }
+
+  handleKeyPress(e) {
+    if (e.nativeEvent.keyCode === 13) {
+      this.applyDirective();
+      this.preventPropagation(e);
+    }
+  }
+
+  columnAlreadyExists(colName) {
+    let headers = DataPrepStore.getState().dataprep.headers;
+    return headers.indexOf(colName) !== -1;
+  }
+
+  switchColumnOrder() {
+    if (this.state.newColumnInput === `${this.state.firstColumn}_${this.state.secondColumn}`) {
+      this.setState({
+        newColumnInput: `${this.state.secondColumn}_${this.state.firstColumn}`
```

Comment:
Do we need two setStates? We could probably combine them using Object.assign and constructing the final object for setState.